lib/connection/connection.js

```diff
@@ -528,97 +528,113 @@ function merge(options1, options2) {
-  connection.setKeepAlive(self.keepAlive, self.keepAliveInitialDelay);
-  connection.setTimeout(self.connectionTimeout);
-  connection.setNoDelay(self.noDelay);
-
-  return connection;
-}
+    // Set default sni servername to be the same as host
+    if (options.servername == null) {
+      options.servername = conn.host;
+    }
 
-function makeUnsecureConnection(self, family) {
-  // Create new connection instance
-  let connection_options;
-  if (self.domainSocket) {
-    connection_options = { path: self.host };
+    options = Object.assign({}, options, { host: conn.host, port: conn.port });
   } else {
-    connection_options = { port: self.port, host: self.host };
-    connection_options.family = family;
+    if (conn.domainSocket) {
+      options = { path: conn.host };
+    } else {
+      options = { port: conn.port, host: conn.host };
+    }
   }
+  return options;
+}
 
-  const connection = net.createConnection(connection_options);
-
-  // Set the options for the connection
-  connection.setKeepAlive(self.keepAlive, self.keepAliveInitialDelay);
-  connection.setTimeout(self.connectionTimeout);
-  connection.setNoDelay(self.noDelay);
+function makeConnection(conn, options, callback) {
+  const module = options.ssl ? tls : net;
 
-  connection.once('connect', function() {
+  const connection = module.connect(options, function() {
+    if (conn.ssl) {
+      // Error on auth or skip
+      if (connection.authorizationError && conn.rejectUnauthorized) {
+        return conn.emit('error', connection.authorizationError, conn, { ssl: true });
+      }
+    }
     // Set socket timeout instead of connection timeout
-    connection.setTimeout(self.socketTimeout);
-    // Emit connect event
-    self.emit('connect', self);
+
+    connection.setTimeout(conn.socketTimeout);
+    return callback(null, connection);
   });
 
-  return connection;
+  // Set the options for the connection
+  connection.setKeepAlive(conn.keepAlive, conn.keepAliveInitialDelay);
+  connection.setTimeout(conn.connectionTimeout);
+  connection.setNoDelay(conn.noDelay);
+
+  // Add handlers for events
+  connection.once('error', err => callback(err, null));
+  connection.once('timeout', err => callback(err, null));
+  connection.once('close', err => callback(err, null));
+  return;
 }
 
-function doConnect(self, family, _options, _errorHandler) {
-  self.connection = self.ssl
-    ? makeSSLConnection(self, _options)
-    : makeUnsecureConnection(self, family);
+function fastFallbackConnect(conn, _options, callback) {
+  const options = prepareConnectionOptions(conn, _options);
+
+  let connection;
+  const connectionHandler = (err, _connection) => {
+    if (_connection) {
+      if (connection) {
+        ['error', 'timeout', 'close'].forEach(event => _connection.removeAllListeners(event));
+        _connection.end();
+        _connection.unref();
+        return;
+      }
+      connection = _connection;
+      return callback(null, connection);
+    } else if (err) {
```

Comment:
@mbroadst 
Three tests are failing because they expect an error to be emitted here with the line

`if (conn.listeners('error').length > 0) conn.emit('error', new MongoNetworkError(err), conn);`
from the errorHandler function.

However, this functionality should only exist when the family is 4 or the family is passed in (which isn't implemented yet), and when I try to implement this for an ipv4 family, I get the issue of done being called twice in the connection tests.

Reply:
Right, we still need to emit an error if one is returned. I think this event handler should look something like this:
```
let errors = [];
let connection;
function connectionHandler(err, _connection) => {
  if (connection) {
    // we already have an active connection, clean this one up
    ['error', 'timeout', 'close'].forEach(event => _connection.removeAllListeners(event));
    _connection.end();
    _connection.unref();
    return;
  }

  if (err) {
    if (errors.length > 0) {
      // an error occurred for the second time, we have officially failed
      return callback(new MongoError('failed to connect'), null);  // maybe make a MongoError an attach both errors to it? 
    }

    // otherwise push the error, and wait for subsequent connects
    errors.push(err);
    return;
  }

  // we have a successful connection, set the connection and callback!
  connection = _connection;
  callback(null, connection);
}
```